Updated output log with multiple enhancements and feature additions.

EricEzaM opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/898#issue-624711321

From Output Log Tracker (#19122):
Closes part of #10870 (no. 2 on that list)
Closes #19098 (which was already closed but not addressed. A similar issue was not recreated in the Proposals repo)
Closes https://github.com/godotengine/godot-proposals/issues/971

Supercedes:
#39274
The above implementation only filters _incoming_ messages. It does not filter messages already in the log.

#38593
The idea of #38593 was to save vertical space. Putting them in a context menu obscures them a bit, so I think putting them vertically is a good compromise. Vertical space is more "valuable" in an output log so this achieves that goal without hiding options.

**Message Type Filters:**
![filters_showcase](https://user-images.githubusercontent.com/41730826/90373858-ed1ca000-e0b5-11ea-8cca-8ce0120087a4.gif)

**Collapse/Stack Duplicate Messages**
![collapse_showcase](https://user-images.githubusercontent.com/41730826/90373891-f9a0f880-e0b5-11ea-8d84-775df7aebf2d.gif)

**Searching**
![search_showcase](https://user-images.githubusercontent.com/41730826/90373904-fe65ac80-e0b5-11ea-923e-f1688b21663b.gif)

There are 2 commits in this PR on purpose. The first adds all log enhancements, the second adds a method to RichTextLabel to improve performance of this output log implementation. **Justification for addition:**

When performing this sequence of operations (which is the main thing the "collapse" functionality uses):

1. remove_line(last_line);
2. add_newline()
3. \<\<Push items as normal\>\>

Extra “newline” items are left behind after the removal. This is because remove_line() removes the content from a line but does not remove the “newline” item on the previous line. This means that if you do several remove_line(), add_newline() operations repeatedly, you get:

- Line1 Content, NewLine
- Line2 Content, NewLine, NewLine, NewLine, NewLine… etc
- Line3 Content

Because there is logic which iterates over all items in remove_item, this data duplication severely slows the speed of execution.
This sequence of operations is critical for the “collapse” functionality of the output log. 

With this change in effect, the log can output, at once (1 frame), 10,000 messages collapsed into one line in **250ms**. Without this change, it takes **12 minutes**.

EricEzaM committed:
Output log rewrite and enhancements.

Added message type filters.
Added ability to search.
Added ability to collapse multiple duplicate messages into one line.
Updated layout to allow for more vertical space in log text area.

EricEzaM committed:
Improve output log performance.

Added method to create a new line in RichTextLabel without adding an ItemNewline to the previous line. Previously, removing a line then adding a newline was adding unnecessary ItemNewline instances to the previous line, significantly the remove_line method.

YeldhamDev commented:
Those four buttons on the top should probably be made flat. To maintain consistency with the rest of the editor.

EricEzaM commented:
@YeldhamDev good point, done

![flatbuttons](https://user-images.githubusercontent.com/41730826/90452818-678b0580-e132-11ea-8ee2-f0eb79f6c103.png)

jordo commented:
I REALLY want this PR to work...  It would be quite usefull.  But I'm having problems understanding why the filter doesn't seem to work correctly:

Base output:
![Screen Shot 2020-10-19 at 4 49 49 PM](https://user-images.githubusercontent.com/323868/96519961-4d9d9a00-122b-11eb-81b3-7fc2639a37d0.png)

Various filter strings:
![Screen Shot 2020-10-19 at 4 50 00 PM](https://user-images.githubusercontent.com/323868/96519987-5b531f80-122b-11eb-9f2d-763b445f01d4.png)
![Screen Shot 2020-10-19 at 4 50 19 PM](https://user-images.githubusercontent.com/323868/96519992-5d1ce300-122b-11eb-9b30-f2039692f486.png)
![Screen Shot 2020-10-19 at 4 50 41 PM](https://user-images.githubusercontent.com/323868/96519999-5ee6a680-122b-11eb-85e9-9892c32fdbd5.png)

It just seems kind of random if the filter string works or not.  Not sure why this would be the case?

EricEzaM commented:
@jordo Hi, thanks for the feedback. That is indeed really weird... The search is done by using `findn(search_text)` so it should only find exact matches of the search string (case insensitive).

I'll see if I can reproduce the issue.

EricEzaM commented:
@jordo I cannot reproduce the issue... It could be the fact that _new_ messages are not filtered when they come are added?

So if you had a filter set like "network" and then "todo" was printed, it would still be added to the log. This is probably a bug and would need to be fixed.

jordo commented:
I'll also clarify my screenshots...  I didn't have any filter enabled when the logs were printed (first screenshot).  All the previous screenshots were filtering the same log as was posted in the first screenshot.

EricEzaM commented:
I just pushed some updates... but I think they are unrelated to your issue. You could try pulling the latest branch and giving it a test.

Otherwise, like I said, I can't seem to reproduce the issue. If you could put a breakpoint at editor_log.cpp on line 167, that is where the filter check happens. Some testing around that area would likely help to pinpoint where the issue is.

jordo commented:
OK.  I'll pull in these changes tomorrow and if i still have issues I'll debug some myself to try and see whats going on.

jordo commented:
I pulled in the latest changes, and the filters still really don't react the way i think they should.  Seems like they still just kind of filter out lines randomly almost


![recording](https://user-images.githubusercontent.com/323868/96615363-67d48800-12be-11eb-82e9-3fdef947106e.gif)

jordo commented:
I guess one thing to note.  This output has interleaved output from a 'play scene' command.  (Which interleaves output in that output window).  It looks like that seems to be the causing the filter errors.

As well, the various filter buttons 'selected' states should probably be saved and loaded to their last selected states per project.  For example, I don't use the 'collapse duplicate messages' button.  But every time I re-open my project it's defaulted to enabled, so I have to remember to turn it off again every time I open my project.  Do the button presses states get saved to project settings?

jordo commented:
Alright, so I believe I found the issue regarding the filter:  Various 'print_line()' calls, can actually get routed into editor_log as one message (with newlines).  Running a scene from the editor actually forks another process, so there's obviously some method to route output back into to the godot editor window.  I'm not sure how the messages are split in these cases, but maybe they come back in segments or bulk or something.  At any rate, my module just uses 'print_line()' a few times, but editor_log.cpp sees it as one message (with two newline characters).

So, given messages can contain newlines, i think the filter should just work on a line-by-line basis like what is common in other IDE's.  Below is an example from VSCode:  The filter for 'thread' 

![Screen Shot 2020-10-20 at 11 25 21 AM](https://user-images.githubusercontent.com/323868/96622491-3ca26680-12c7-11eb-9f18-85fa7e417727.png)

![Screen Shot 2020-10-20 at 11 25 31 AM](https://user-images.githubusercontent.com/323868/96622502-40ce8400-12c7-11eb-96fc-225a502f225c.png)

jordo commented:
@EricEzaM  I've fixed my newline/filtering issue here:  https://github.com/winterpixelgames/godot/commit/fdebfb36a891ad7c724e4b040914a19bc38d966a  if you would like to bring in.

jordo commented:
Only other suggestion I would make, is to by default NOT collapse duplicate messages ( defaulting to the existing functionality).  And, by default, select/show the search box (eliminates the need for users to explicitly try and find this functionality).  The search filter is almost by default enabled on every other IDE I've ever used.

jordo commented:
@EricEzaM I implemented the above suggestion here:  https://github.com/winterpixelgames/godot/commit/9ab407474576ef0df0ef5ccbf817c56dd260ad6c  As well, I disabled focus on  the right-side buttons, which is better from a usability perspective.  Clicking a button was taking focus away from others (say the search box).  These buttons don't really ever need to be in focus as they are really just toggles.

EricEzaM commented:
@jordo thanks for that! So the issues at it's source is that multiple `print()` statements can be grouped together into one editorlog call... hmm... Maybe that should be changed rather than working around it? Might need some input for others here, but I'm not sure who. @akien-mga could you help here or let us know who to ask?

jordo commented:
Well, ya, I guess two issues really.  

1) - the buttons focus, (which should be addressed in my second attached commit).  and

2) - Input filtering should be line-by-line like grep.  I don't think it's really a workaround, because that's really how the filter should work.  There's no difference in the output window between:  `print("hi\nthere")` and `print("hi"); print("there")`, it looks the same.  So when a user types 'hi' into the filter box, they should expect to see the same result, regardless of the underlying messages in editorlog.

CitrusWire commented:
This PR looks great, but I do have a question/query. Noting I've only seen the screenshots above and am still new to Godot so maybe I'm missing something:
Why is this changing stuff on the "Output" panel, rather than the "Debugger > Errors" panel? For me at least currently if Godot has a "Warning", it punts it to the Debugger>Errors, and there's no sign of it on Output. Could/should this not also be carried over to the Debugger>Errors? (Or maybe it is and you just don't show it in the screenshots, in which case ignore me :-) )

EricEzaM commented:
@CitrusWire yes errors and warnings are also shown in the output log, but not all of them it seems.

CitrusWire commented:
> @CitrusWire yes errors and warnings are also shown in the output log, but not all of them it seems.

@EricEzaM - Thanks for the clarification. I appreciate it's unrelated to this, but wouldn't that be a bug? Is it worth opening a bug report for it? I can't see one for it presently.

More ontopic, my suggestion for this PR would be to carry this over to the Debugger too, though I have no idea how feasible that is. This would provide more consistency for the users.

menip commented:
This PR includes collapsing/stacking, but not grouping from what I can tell. I describe my idea in detail here: https://github.com/godotengine/godot-proposals/issues/971#issuecomment-794801707

akien-mga commented on editor/editor_log.cpp:
```diff
@@ -94,15 +113,70 @@ void EditorLog::clear() {
-void EditorLog::copy() {
-	_copy_request();
+void EditorLog::add_message(const String &p_msg, MessageType p_type) {
+	if (messages.size() > 0 && messages[messages.size() - 1].text == p_msg) {
+		// If previous message is the same as the new one, increase previous count rather than adding another
+		// instance to the messages list.
+		LogMessage &previous = messages.write[messages.size() - 1];
+		previous.count++;
```
Errors can often get repeated but not contiguous, e.g.:
```
ERROR: get_action_strength: Request for nonexistent InputMap action 'look_right'.
   At: main/input_default.cpp:142.
ERROR: get_action_strength: Request for nonexistent InputMap action 'look_left'.
   At: main/input_default.cpp:142.
ERROR: get_action_strength: Request for nonexistent InputMap action 'look_down'.
   At: main/input_default.cpp:142.
ERROR: get_action_strength: Request for nonexistent InputMap action 'look_up'.
   At: main/input_default.cpp:142.
ERROR: get_action_strength: Request for nonexistent InputMap action 'look_right'.
   At: main/input_default.cpp:142.
ERROR: get_action_strength: Request for nonexistent InputMap action 'look_left'.
   At: main/input_default.cpp:142.
ERROR: get_action_strength: Request for nonexistent InputMap action 'look_down'.
   At: main/input_default.cpp:142.
ERROR: get_action_strength: Request for nonexistent InputMap action 'look_up'.
   At: main/input_default.cpp:142.
ERROR: get_action_strength: Request for nonexistent InputMap action 'look_right'.
   At: main/input_default.cpp:142.
ERROR: get_action_strength: Request for nonexistent InputMap action 'look_left'.
   At: main/input_default.cpp:142.
ERROR: get_action_strength: Request for nonexistent InputMap action 'look_down'.
   At: main/input_default.cpp:142.
ERROR: get_action_strength: Request for nonexistent InputMap action 'look_up'.
   At: main/input_default.cpp:142.
```

Typically happens when triggering multiple errors in a given frame which are then repeated each frame. It can also happen with chained errors when the engine prints several error messages for a single fault (e.g. when a scene can't be loaded due to missing imported dependencies).

Is this something we could handle? That would require storing all messages in a hashmap to increment the counter regardless of the position, that might impact performance I guess.

Alternatively this could maybe be done over e.g. the last 10 messages, so it would group related duplicate messages which happen in close sequence, but not a later similar message spam from another source. (E.g. physics errors triggered by two different physics body that the player interacts with at different times.)

EricEzaM replied:
We could look into this I guess but I am hesitant for now, as displaying them "grouped" may mess with the order in which they occur.

akien-mga commented on editor/editor_log.cpp:
```diff
@@ -125,20 +199,42 @@ void EditorLog::add_message(const String &p_msg, MessageType p_type) {
 
-void EditorLog::set_tool_button(Button *p_tool_button) {
-	tool_button = p_tool_button;
+void EditorLog::_set_filter_active(bool p_active, int p_message_type) {
```
This should be a `MessageType` parameter no?

akien-mga commented on editor/editor_log.h:
```diff
@@ -36,19 +36,92 @@
+		Button *toggle_button = nullptr;
+		bool active = true;
+
+		void initialise_button(const String &p_tooltip, Callable p_toggled_callback) {
```
Sorry Commonwealth, the US won the IT wars.
```suggestion
		void initialize_button(const String &p_tooltip, Callable p_toggled_callback) {
```

akien-mga replied:
Is there a specific reason to implement this in the header instead of the `.cpp`?

akien-mga commented on editor/editor_log.cpp:
```diff
@@ -62,6 +62,18 @@ void EditorLog::_notification(int p_what) {
 		log->add_theme_color_override("selection_color", get_theme_color("accent_color", "Editor") * Color(1, 1, 1, 0.4));
+		log->add_theme_font_override("bold_font", get_theme_font("bold", "EditorFonts"));
+
+		type_filter_map.find(MSG_TYPE_STD)->get()->toggle_button->set_icon(get_theme_icon("Popup", "EditorIcons"));
+		type_filter_map.find(MSG_TYPE_ERROR)->get()->toggle_button->set_icon(get_theme_icon("StatusError", "EditorIcons"));
+		type_filter_map.find(MSG_TYPE_WARNING)->get()->toggle_button->set_icon(get_theme_icon("StatusWarning", "EditorIcons"));
+		type_filter_map.find(MSG_TYPE_EDITOR)->get()->toggle_button->set_icon(get_theme_icon("Edit", "EditorIcons"));
```
If we're sure that those four entries exist, maybe they can be indexed directly as done in other places?
```suggestion
		type_filter_map[MSG_TYPE_STD]->toggle_button->set_icon(get_theme_icon("Popup", "EditorIcons"));
		type_filter_map[MSG_TYPE_ERROR]->toggle_button->set_icon(get_theme_icon("StatusError", "EditorIcons"));
		type_filter_map[MSG_TYPE_WARNING]->toggle_button->set_icon(get_theme_icon("StatusWarning", "EditorIcons"));
		type_filter_map[MSG_TYPE_EDITOR]->toggle_button->set_icon(get_theme_icon("Edit", "EditorIcons"));
```

akien-mga reviewed (approved):
Left some comments but looks great overall. This is a big UX improvement compared to the current situation which has many users completely ignoring errors being reported in the debugger.

jordo commented:
Please be aware that newlines do not filter correctly without https://github.com/godotengine/godot/commit/fdebfb36a891ad7c724e4b040914a19bc38d966a

EricEzaM commented:
Hi Akien and Jordo,

Thanks for the feedback and help! I will look at reading this with jordos improvements included.

EricEzaM commented:
With the filtering issue @jordo mentioned, I am not sure if splitting by newline is the best option. Then errors/prints which legitimately use new lines would not be "collapsed" correctly, as each line would be treated as a new entry.

If possible, I think it may be better to resolve this at the source - where multiple messages are being passed as one?

jordo commented:
> If possible, I think it may be better to resolve this at the source - where multiple messages are being passed as one to EditorLog?

This isn't going to work with messages coming into the editor from a forked process unless stdout is flushed after every newline, which probably just isn't going to happen.   

Without this, filtering is basically pointless as you'll get something like below, which will be very counter intuitive for just about everybody:
![image](https://user-images.githubusercontent.com/323868/115818477-0f3f8e00-a3ba-11eb-8353-0867aad74b8d.png)

It does impact how collapsing works wrt messages that may include newlines themselves, but this is how filters work on just about every other IDE out there.

EricEzaM commented:
Updated. Thanks @jordo for the help with testing,

For the splitting by the new lines I think it can be done using the `split` method already part of the `String` class
```cpp
Vector<String> lines = p_msg.split("\n", false);

	for (int i = 0; i < lines.size(); i++) {
		_process_message(lines[i], p_type);
	}
}
```

akien-mga commented:
Thanks!

dalexeev commented:
@EricEzaM Could you please remove the blank line at the beginning? You probably didn't notice [this PR](https://github.com/godotengine/godot/pull/44909) when rebasing.

![](https://user-images.githubusercontent.com/47700418/117207934-36a44b00-adfd-11eb-870c-b7f45a97d3fd.png)
